### PR TITLE
feat: add hashicorp vault transit to encrypt and decrypt asset password

### LINKED ISSUE #80: feature: use HashiCorp Vault to store asset password
If we use a password vault to store the username and password of the servers, the security of the service will increase

### PR DESCRIPTION
**Pull request by tulutulupi**
Fix: https://github.com/casvisor/casvisor/issues/80

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,8 @@
 module github.com/casvisor/casvisor
 
-go 1.20
+go 1.21
```
Comment (hsluoyz) [inline go.mod]: Don't upgrade the Go version. You can use Go 1.20.x in your dev env

File: util/vaultService.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		return nil, fmt.Errorf("failed to create vault client: %v", err)
+	}
+
+	client.SetToken(beego.AppConfig.String("vaultToken"))
+
+	return client, nil
+}
+
+func EncryPassword(client *vault.Client, plaintext string) (string, error) {
+	// base64 编码
```
Comment (hsluoyz) [inline util/vaultService.go]: All English

File: util/vaultService.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	secret, err := client.Logical().Write("transit/decrypt/my-key", data)
+	if err != nil {
+		return "", fmt.Errorf("failed to decrypt data: %v", err)
+	}
+
+	plaintextBase64 := secret.Data["plaintext"].(string)
+
+	// base64 解码
```
Comment (hsluoyz) [inline util/vaultService.go]: All English

File: object/asset.go [hunk 4]
```diff
@@ -138,6 +138,13 @@ func getAsset(owner string, name string) (*Asset, error) {
 	}
 
 	if existed {
+		vaultClient, err := util.InitVaultClient()
+		if err == nil {
+			asset.Password, err = util.DecryPassword(vaultClient, asset.Password)
```
Comment (hsluoyz) [inline object/asset.go]: typo, Decrypt

File: object/asset.go [hunk 5]
```diff
@@ -138,6 +138,13 @@ func getAsset(owner string, name string) (*Asset, error) {
 	}
 
 	if existed {
+		vaultClient, err := util.InitVaultClient()
```
Comment (hsluoyz) [inline object/asset.go]: Why do you handle get? why not connect (via RDP/SSH, etc.)?

File: object/asset.go [hunk 6]
```diff
@@ -204,6 +211,13 @@ func UpdateAsset(id string, asset *Asset) (bool, error) {
 		asset.SshStatus = AssetStatusStopped
 	}
 
+	vaultClient, err := util.InitVaultClient()
+	if err == nil {
+		asset.Password, err = util.EncryPassword(vaultClient, asset.Password)
```
Comment (hsluoyz) [inline object/asset.go]: Define `getMaskedAsset` like: https://github.com/casdoor/casdoor/blob/4a3bd84f848fa3756b0510d80fd148e00e2b8dfa/controllers/user.go#L234

File: util/vaultService.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
+// Copyright 2023 The casbin Authors. All Rights Reserved.
```
Comment (hsluoyz) [inline util/vaultService.go]: put this new file to new "vault" pkg

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: go.mod [hunk 8]
```diff
@@ -15,43 +17,75 @@ require (
...
+	github.com/gofrs/uuid v4.4.0+incompatible // indirect
 	github.com/golang-jwt/jwt/v4 v4.5.0 // indirect
 	github.com/golang/protobuf v1.5.3 // indirect
 	github.com/golang/snappy v0.0.4 // indirect
 	github.com/gomodule/redigo v2.0.0+incompatible // indirect
 	github.com/google/go-cmp v0.6.0 // indirect
+	github.com/hashicorp/errwrap v1.1.0 // indirect
+	github.com/hashicorp/go-cleanhttp v0.5.2 // indirect
+	github.com/hashicorp/go-multierror v1.1.1 // indirect
+	github.com/hashicorp/go-retryablehttp v0.7.6 // indirect
```
Comment (hsluoyz) [inline go.mod]: Can we avoid importing so many hashicorp libs? Just copy-paste the HTTP API code we needed from its SDK

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: object/asset.go (hunk 5 above)
Comment (tulutulupi) [reply object/asset.go]: I had a discussion with Supervisor Yu. He added the mask in the asset.go file within the controller. So, I first decrypt the data, then add the mask. The password will not be exposed to the front-end.

File: object/asset.go [hunk 9]
```diff
@@ -205,6 +207,8 @@ func UpdateAsset(id string, asset *Asset) (bool, error) {
 		asset.SshStatus = AssetStatusStopped
 	}
 
+	asset.Password, _ = vault.GetEncryptedPassword(asset.Password)
```
Comment (hsluoyz) [inline object/asset.go]: All errors should be handled

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**